common/lib/plugins/failover/failover_plugin.ts

```diff
@@ -441,14 +433,28 @@ export class FailoverPlugin extends AbstractConnectionPlugin {
             throw new AwsWrapperError(Messages.get("Failover.unableToDetermineWriter"));
           }
 
+          const allowedHosts = this.pluginService.getHosts();
+          if (!allowedHosts.some((hostInfo: HostInfo) => hostInfo.host === writerHostInfo?.host)) {
+            const failoverErrorMessage = Messages.get(
+              "Failover.newWriterNotAllowed",
+              writerHostInfo ? writerHostInfo.host : "<null>",
+              logTopology(allowedHosts, "[Failover.newWriterNotAllowed]")
+            );
+            logger.error(failoverErrorMessage);
+            await this.pluginService.abortTargetClient(result.client);
+            throw new FailoverFailedError(failoverErrorMessage);
+          }
+
           await this.pluginService.abortCurrentClient();
           await this.pluginService.setCurrentClient(result.client, writerHostInfo);
           logger.debug(Messages.get("Failover.establishedConnection", this.pluginService.getCurrentHostInfo()?.host ?? ""));
           await this.pluginService.refreshHostList();
-          this.throwFailoverSuccessError();
           this.failoverWriterSuccessCounter.inc();
+          this.throwFailoverSuccessError();
         } catch (error: any) {
-          this.failoverWriterFailedCounter.inc();
+          if (!(error instanceof FailoverSuccessError)) {
+            this.failoverWriterFailedCounter.inc();
```

Comment:
`throwFailoverSuccessError` might throw `TransactionResolutionUnknownError`, so its better to call `this.failoverWriterSuccessCounter.inc();` in the catch block